### DISCUSSION TITLE
422 unprocessable entity (field_required)

### DISCUSSION THREAD

**Discussion by hgue** [pydantic/pydantic#4975] [General]
Hello,
sorry for asking this here, but I am totally out of ideas with my case regarding testing the fastapi api using the TestClient and pydantic for data checking. Hope someone can lead me in the right direction.

Please look at the following code:

####################################################################

# -> model

from typing import Optional
from datetime import datetime
from pydantic import BaseModel, validator

class CoreModel(BaseModel):
"""
Any common logic to be shared by all models goes here
"""
    pass

class UserCreate(CoreModel):
"""
Email, username, and password are required for registering a new user
"""
    email: EmailStr
    password: constr(min_length=7, max_length=100)
    username: constr(min_length=3, regex="[a-zA-Z0-9_-]+$")

class Config:
    # The lineorm_mode = True allows the app to take ORM objects and translate
    # them into responses automatically. This automation saves us from manually
    # taking data out of ORM, making it into a dictionary, then loading it in with Pydantic.
    orm_mode = True

# -> api route

@router.post("/", status_code=status.HTTP_201_CREATED, name="users:register-new-user")
def register_new_user(
    new_user: UserCreate = Body(..., embed=True),
    users_service: UsersService = Depends(get_service(UsersService, UsersRepository))
) -> UserPublic:

    result = users_service.register_new_user(new_user)
    created_user = result.handle()
    ...

# -> test logic (pytest)

@pytest.fixture
def user_to_create():
return {"new_user": {"email": "[shakira@shakira.io](mailto:shakira@shakira.io)", "password": "chantaje", "username": "shakirashakira"}}

def test_users_can_register_successfully(client: TestClient, user_to_create):
    users_service = UsersService(client.db, UsersRepository)
    res = client.post(client.app.url_path_for("users:register-new-user"), json=user_to_create)
    print(user_to_create)
    print(res.json())
    
    assert res.status_code == status.HTTP_201_CREATED
####################################################################

The last line throws the following exception when using pytest:

2023-01-21 09:46:27 | httpx._client | _client | HTTP Request: POST http://testserver/api/v1/users/ "HTTP/1.1 422 Unprocessable Entity"
{'new_user': {'email': '[shakira@shakira.io](mailto:shakira@shakira.io)', 'password': 'chantaje', 'username': 'shakirashakira'}}
{'detail': [{'loc': ['query', 'kw'], 'msg': 'field required', 'type': 'value_error.missing'}]}
FAILED

Somehow it is not process the 'new_user' object correctly. I am very sure, the test data is correct, it is working with the swagger documentation on the website using the endpoint with the same dataset (201). I don't understand what "field required" is missing here? The values are all supplied in this case..

Any ideas howto debug this?
Thank you a lot for your support!

**Comment by samuelcolvin** [comment]
At the very least you need to format your code properly, but my instinct is that this isn't something we can help with.

**Comment by ezeparziale** [comment]
Hello!

Check your docs in `http://127.0.0.1:8000/docs`

I simplicity your code to check. 
The test run with `pytest -v -s` and I have checked and it all looks good.


`main.py`
```
from fastapi import FastAPI, status, Body
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, validator, EmailStr, constr

app = FastAPI()


class CoreModel(BaseModel):
    """
    Any common logic to be shared by all models goes here
    """
    pass


class UserCreate(CoreModel):
    """
    Email, username, and password are required for registering a new user
    """
    email: EmailStr
    password: str
    username: str

    class Config:
        # The lineorm_mode = True allows the app to take ORM objects and translate
        # them into responses automatically. This automation saves us from manually
        # taking data out of ORM, making it into a dictionary, then loading it in with Pydantic.
        orm_mode = True


class UserPublic(CoreModel):
    email: str
    username: str


@app.post("/", status_code=status.HTTP_201_CREATED)
def register_new_user(
    new_user: UserCreate,
) -> UserPublic:
    data = UserPublic(email=new_user.email, username=new_user.username)
    return data
```

`test_post.py`
```
from fastapi.testclient import TestClient

from .main import app

client = TestClient(app)


def test_post():
    res = client.post(
        "/",
        json={"email": "user@example.com", "password": "string", "username": "string"},
    )

    print(res.text)
    assert res.status_code == 201

```


Check my repo with a full example: https://github.com/ezeparziale/fastapi-auth-google-template

**Comment by ryanjdillon** [comment]
I got this error using the FastApi CrudRouter w/ SqlAlchemy, as I passed a scoped session object to the `db` param, passing a LocalSession did the trick.
